### PR TITLE
PyPI Analysis: enhance analysis of code and dep files

### LINKED ISSUE #402: Distinguish between notebooks and Python files in the source code

### PR DESCRIPTION
**Pull request by zz1874**
Closes #402 and #403

This PR has a series of updates to enhance the analysis and results generation process. 

Notable improvements include:
- for code directory analysis
  - split code directories detection into `.py` files and `.ipynb` files
  - count the number of `.py` files, `.ipynb` files and the total number of Python files for each directory
- for dependency declaration files analysis
  - count the number of dependencies declared in each dependency declaration file
  - add `warnings` attribute to mark whether there's a warning while parsing this dependency declaration file
- for import patterns detection
  - rename `imports` to `detected_imports` in `Analysis()` to avoid misunderstanding
- results analysis
  - update the JSON results
  - update the data visualization
    - update the Venn diagram with an explicit legend

### CODE REVIEW

File: PyPI_analysis/main.py [hunk 1]
```diff
@@ -292,6 +359,14 @@ def main(
         return 5
 
     exit_code = assign_exit_code(analysis=analysis)
-    print_output(analysis=analysis, stdout=stdout)
+
+    log_contents = log_buffer.getvalue()
+    warning_files = {
```
Comment (mknorps) [inline PyPI_analysis/main.py:364]: I think we could change the name of this to `warnings_entries` or `warnigs_limited_eval`. Current name suggests that there are files involved.

**Review by mknorps — APPROVED** [review]
Great job @Zhihan!

You handled well the warnings count and expanding information on code and dependencies source.

I have some minor, non-blocking comments.

File: PyPI_analysis/main.py [hunk 2]
```diff
@@ -99,8 +104,36 @@ def imports(self) -> List[ParsedImport]:
...
+            extract_declared_dependencies.parse_sources(
+                (src for src in self.sources if isinstance(src, DepsSource))
+            )
+        )
+
+    @property
+    @calculated_once
+    def dep_files(self) -> Dict[DepsSource, int]:
+        """The dictionary of dependency declaration files and dependency count"""
+        dep_sources = {src for src in self._sources if isinstance(src, DepsSource)}
```
[the remark this answers is not in the document]
Comment (zz1874) [reply PyPI_analysis/main.py]: Addressed in https://github.com/tweag/FawltyDeps/pull/404/commits/b07989f82b3924ec305dc5064d3e1be9e0bdcf6b

File: PyPI_analysis/main.py (hunk 1 above)
Comment (zz1874) [reply PyPI_analysis/main.py]: Addressed in https://github.com/tweag/FawltyDeps/pull/404/commits/b07989f82b3924ec305dc5064d3e1be9e0bdcf6b

**Pull request merged**